src/metatrain/utils/llpr.py

```diff
@@ -248,10 +254,91 @@ class in ``metatrain``.
             )
             output = self.model(systems, options, check_consistency=False)
             ll_feat_tmap = output["mtt::aux::last_layer_features"]
-            ll_feats = ll_feat_tmap.block().values / n_atoms.unsqueeze(1)
+            ll_feats = ll_feat_tmap.block().values.detach() / n_atoms.unsqueeze(1)
             self.covariance += ll_feats.T @ ll_feats
         self.covariance_computed = True
 
+    def compute_covariance_as_pseudo_hessian(
+        self,
+        train_loader: DataLoader,
+        target_infos: TargetInfoDict,
+        loss_fn: Callable,
+        parameters: List[torch.nn.Parameter],
+    ) -> None:
+        """A function to compute the covariance matrix for a training set
+        as the pseudo-Hessian of the loss function.
+
+        The covariance/pseudo-Hessian is stored as a buffer in the model. The
+        loss function must be compatible with the Dataloader (i.e., it should
+        have the same structure as the outputs of the model and as the targets
+        in the dataset). All contributions to the loss functions are assumed to
+        be per-atom, except for quantities that are already per-atom (e.g.,
+        forces).
+
+        :param train_loader: A PyTorch DataLoader with the training data.
+            The individual samples need to be compatible with the ``Dataset``
+            class in ``metatrain``.
+        :param loss_fn: A loss function that takes the model outputs and the
+            targets and returns a scalar loss.
+        :param parameters: A list of model parameters for which the pseudo-Hessian
+            should be computed.
+        """
+        self.model = self.model.train()  # we need gradients w.r.t. parameters
+        # disable gradients for all parameters that are not in the list
+        for parameter in self.model.parameters():
```

Comment:
The logic behind the introduction of `parameters` is very unclear. What are the parameters? How does the relate to the last-layer features? How are these parameters to be defined, and then extracted for the covariance matrix to be correctly calculated? This _needs_ to be documented.